Sources/Plasma/Apps/plClient/Mac-Cocoa/plMacDisplayHelper.mm

```diff
@@ -0,0 +1,147 @@
+    // depending on dock and menu size so I'll return that instead.
+    NSRect windowArea = [screen visibleFrame];
+    fDesktopDisplayMode.Width = windowArea.size.width;
+    fDesktopDisplayMode.Height = windowArea.size.height;
+    fDesktopDisplayMode.ColorDepth = 32;
+    
+    // visibleFrame is in points - put into pixels
+    fDesktopDisplayMode.Width *= [screen backingScaleFactor];
+    fDesktopDisplayMode.Height *= [screen backingScaleFactor];
+    
+    // Calculate the region actually available for full screen
+    NSRect currentResolution = [screen frame];
+    if (@available(macOS 12.0, *)) {
+        NSEdgeInsets currentSafeAreaInsets = [screen safeAreaInsets];
+        // Sigh... Origin doesn't matter but lets do it for inspectability
+        currentResolution.origin.x += currentSafeAreaInsets.left;
+        currentResolution.origin.y += currentSafeAreaInsets.top;
+        currentResolution.size.width -= currentSafeAreaInsets.left + currentSafeAreaInsets.right;
+        currentResolution.size.height -= currentSafeAreaInsets.top + currentSafeAreaInsets.bottom;
+    }
+
+    float safeAspectRatio = currentResolution.size.width / currentResolution.size.height;
+
+    fDisplayModes.clear();
+
+    CFArrayRef displayModes = CGDisplayCopyAllDisplayModes(fCurrentDisplay, nullptr);
+    for(int i=0; i< CFArrayGetCount(displayModes); i++)
+    {
+        // Now filter out the ones that are taller than the safe area aspect ratio
+        // This could break in interesting ways if Apple ships displays that have unsafe
+        // areas along the side - but will prevent us stripping any aspect ratios that don't
+        // match the display entirely.
+        //
+        // I'd prefer not to filter a bunch of stuff out but don't strictly have a choice
+        // here because there are a lot of fake resolutions in the list.
+        //
+        // Apple also doesn't include some traditional resolutions on some of their hardware
+        // like 800x600, so don't expect those here. My Macbook Pro gives me back 1080p as the lowest.
+        // We could force that with kCGDisplayShowDuplicateLowResolutionModes but the aspect ratio
+        // filter would still drop some or all of those.
+        //
+        // Asked for better guidance here from Apple - FB13375033
+        
+        CGDisplayModeRef mode = (CGDisplayModeRef)CFArrayGetValueAtIndex(displayModes, i);
+
+        float modeAspectRatio = float(CGDisplayModeGetWidth(mode)) / float(CGDisplayModeGetHeight(mode));
+        if(modeAspectRatio < safeAspectRatio)
+        {
+            continue;
+        }
+
+        // aspect ratio is good - add the mode to the list
+        plDisplayMode plasmaMode;
+        plasmaMode.ColorDepth = 32;
+        plasmaMode.Width = int(CGDisplayModeGetWidth(mode));
+        plasmaMode.Height = int(CGDisplayModeGetHeight(mode));
+
+        fDisplayModes.push_back(plasmaMode);
```

Comment:
Prefer to construct in place with `emplace_back()`.

Reply:
`plDisplayMode` is basically a struct with no customization, so it should support "aggregate initialization" or whatever the default is called. Does `fDisplayModes.emplace_back(width, height, depth);` not work?

Reply:
It does not:
`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.sdk/usr/include/c++/v1/__memory/allocator.h:165:24: error: no matching constructor for initialization of 'plDisplayMode'
  165 |     ::new ((void*)__p) _Up(std::forward<_Args>(__args)...);`
  I was going to remark that plDisplayMode is a class and not a structure - but then I read C++17 should potentially support aggregate initialization from POD type classes. So now I'm not sure why this isn't working.

Reply:
IME you need to have a constructor for construct in place to work, so we should definitely add one. What does work with no extra work is something like `fContainer.emplace_back({ member1, member2, member3 })`, but that may cause copying on these structures with only trivially copyable members.

Reply:
@Hoikas - It feels like that should work (the approach with {}) - but that also doesn't. Not sure why.

Regardless, it sounds like we might want to add a constructor. But I also think it's a little weird it's not a struct? This type originated from Cyan, right?

Reply:
I got it figure out. The syntax with explace is just a little twitchy - different than directly constructing.

Reply:
By the way, a `struct` in C++ behaves exactly like a `class`, except that `struct` defaults to `public` access instead of `private`. So whether something is a `class` with everything public or just a `struct` should never make a difference.